articles/node_lifecycle.md

```diff
@@ -0,0 +1,287 @@
+---
+layout: default
+title: Managed nodes
+abstract:
+  This article describes the concept of a node with a managed life cycle. It
+  aims to document some of the options for supporting managed-life cycle nodes
+  in ROS 2. It has been written with consideration for the existing design of
+  the ROS 2 C++ client library, and in particular the current design of
+  executors.
+author: '[Geoffrey Biggs](https://github.com/gbiggs)'
+published: true
+---
+
+{:toc}
+
+# {{ page.title }}
+
+<div class="abstract" markdown="1">
+{{ page.abstract }}
+</div>
+
+Original Author: {{ page.author }}
+
+## Background
+
+A managed life cycle for nodes allows greater precision over the state of ROS
+system. It will allow roslaunch to have greater control over launching a ROS
+system, and in particular will allow roslaunch to ensure that all components
+have been instantiated correctly before it allows any component to begin
+executing its behaviour. It will also allow nodes to be restarted or replaced
+on-line.
+
+The most important concept of this document is that a managed node presents a
+well-known interface, executes according to a well-known life cycle state
+machine, and otherwise can be considered a black box. This allows freedom to
+the node developer on how they provide the managed life cycle functionality,
+while also ensuring that any tools created for managing nodes can work with any
+compliant node.
+
+
+## Life cycle
+
+![The proposed node life cycle state machine](/img/node_lifecycle/life_cycle_sm.png "The proposed node life cycle state machine")
+
+The behaviour of each state is as defined below.
+
+### Created state
+
+This is the life cycle state the node is in immediately after being
+instantiated. The node may transition to the `inactive` state by being
+initialised.
+
+Implementation note: In object-oriented languages, the onCreated method can be
+mapped to the class constructor.
+
+### Inactive state
+
+This state represents a node that is not currently performing any processing.
+While in this state, the node will not receive any execution time to read
+topics, perform processing of data, respond to service requests, etc. Any data
+that arrives on topics will not be read. Any service requests will not be
+answered.
+
+[To discuss: All of the above, and additionally should data/service requests be
```

Comment:
I'm not very inclined towards queuing data/service requests while the node is not active. At least not make it a default behavior, as I think it can cause more harm than good.

Consider for instance a node that queues a request and gets activated long after, at a time in which the request no longer makes sense, and can potentially cause undesired behavior. To prevent this, the burden is on the implementer to check request timestamps and other meaningful data, and to explicitly reject invalid ones. Although this might seem OK for a power user, I think it's too much to ask to everybody.

So, -1 for making it the default, and +0 for it being a mechanism that can be explicitly enabled. Maybe going over some examples can justify the added complexity and turn this into a +1.